### PR TITLE
Send TestEvent to RS outside transaction

### PR DESCRIPTION
**Pull request by mpbrown**
# BACKEND PULL REQUEST

## Related Issue

- #6209

## Changes Proposed

- Previously, a TestEvent was sent to RS within the same transaction that the TestEvent was saved to the database in. If sending to RS failed such as during FHIR conversion, then the entire transaction would be rolled back.
- Reporting the TestEvent has now been moved outside the transaction into a [TransactionalEventListener](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html). The transactions in `TestOrderService` will now publish an event with the `TestEvent` data and Spring will look for an event listener matching that event type. Currently, Spring will wait to handle that event listener until the transaction successfully commits under the assumption that if there is an error saving the TestEvent to the database then it probably should not be submitted to RS yet.

## Additional Information

- The `testResultsDeliveryService` might pose a similar issue. Although it appears that the `addMultiplexResult` transaction can handle a failed delivery status without causing a rollback, it is possible an uncaught exception from `EmailService` or `SmsService` could rollback that whole transaction.

## Testing

- Verify that TestEvents and TestOrders are both still processed as expected

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java [hunk 1]
```diff
@@ -361,17 +363,10 @@ public AddTestResultResponse addMultiplexResult(
...
-    if (savedEvent.hasCovidResult()) {
-      _testEventReportingService.report(savedEvent);
-    }
 
-    if (savedEvent.hasFluResult() && fhirReportingEnabled) {
-      _fhirQueueReportingService.report(savedEvent);
-    }
+    AddTestResultResponse response = new AddTestResultResponse(savedOrder, deliveryStatus);
+    applicationEventPublisher.publishEvent(new ReportTestEventToRSEvent(savedOrder.getTestEvent()));
+    return response;
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:369]: do we need the `response` variable? can we just return without declaring it?

**Review by zdeveloper — COMMENTED** [review]
Implementation looks great! Will deploy it to dev7 for further testing.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java (hunk 1 above)
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:369]: Oh good catch I forgot to clean that one up when refactoring it

**Review by zdeveloper — APPROVED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ReportTestEventToRSEventListener.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Allows a test event to be sent to ReportStream after a transaction successfully commits. This
+ * ensures that if a failure occurs when sending the test event to ReportStream, any database
+ * changes from the transaction are not rolled back.
+ */
+@Component
+@RequiredArgsConstructor
+public class ReportTestEventToRSEventListener {
+  @Qualifier("csvQueueReportingService")
```
Comment (nathancrtr) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ReportTestEventToRSEventListener.java:21]: TIL about `@Qualifier`...

**Review by nathancrtr — APPROVED** [review]
LGTM! Very keen Java-ing here, I definitely learned a few things.

If it's easy to set up the relevant scenario in-app, it would be nice to see the the created TestEvent row in Metabase after a submission failure, but I feel comfortable with the provided unit tests - it's not a blocker.

**Pull request merged**